packages/jest-runtime/src/index.ts

```diff
@@ -352,16 +352,38 @@ class Runtime {
         importModuleDynamically: (
           specifier: string,
           referencingModule: VMModule,
-        ) =>
-          this.loadEsmModule(
-            this._resolveModule(referencingModule.identifier, specifier),
-          ),
+        ) => {
+          const resolved = this._resolveModule(
+            referencingModule.identifier,
+            specifier,
+          );
+          if (
+            this._resolver.isCoreModule(resolved) ||
+            this.unstable_shouldLoadAsEsm(resolved)
+          ) {
+            return this.loadEsmModule(resolved);
+          }
+
+          return this.loadCjsAsEsm(
+            referencingModule.identifier,
+            resolved,
+            context,
+          );
+        },
         initializeImportMeta(meta: ImportMeta) {
           meta.url = pathToFileURL(modulePath).href;
         },
       });
 
       this._esmoduleRegistry.set(cacheKey, module);
+
+      await module.link((specifier: string, referencingModule: VMModule) =>
```

Comment:
actually quite the bugfix moving this in here, dynamic import of ESM didn't work unless it was cached already 😀